include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,808 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2016 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__quadrature_point_data_h
+#define dealii__quadrature_point_data_h
+
+#ifdef DEAL_II_WITH_CXX11
+
+#include <deal.II/base/quadrature.h>
+#include <deal.II/base/config.h>
+#include <deal.II/base/subscriptor.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/tria_accessor.h>
+#include <deal.II/distributed/tria.h>
+#include <deal.II/fe/fe_tools.h>
+#include <deal.II/fe/fe.h>
+#include <deal.II/lac/vector.h>
+
+#include <vector>
+#include <map>
+#include <typeinfo>
+
+DEAL_II_NAMESPACE_OPEN
+
+/*!@addtogroup Quadrature */
+/*@{*/
+
+/**
+ * A class for storing at each cell @p CellIteratorType a vector of data @p DataType .
+ * The underlying structure and the initialize() method of this class is designed
+ * in such a way that one could use different child classes derived from the
+ * base DataType class to store data on a given cell. This implies the usage of pointers,
+ * in our case -- std::shared_ptr().
+ *
+ * @note Note that the data type stored on each cell can be different.
+ * However, within the cell this class stores a vector of objects of a single data type.
+ * For this reason, this class may not be sufficiently flexible when, for example,
+ * adopting a level-set approach to describe material behavior.
+ *
+ * @author Denis Davydov, Jean-Paul Pelteret, 2016
+ */
+template <typename CellIteratorType, typename DataType>
+class CellDataStorage
+{
+public:
+  /**
+   * A typedef for a vector of shared pointers with arbitrary data type @p T .
+   */
+  template <typename T> using CellVector = std::vector<std::shared_ptr<T> >;
+
+  /**
+   * Default constructor.
+   */
+  CellDataStorage();
+
+  /**
+   * Default virtual destructor.
+   */
+  virtual ~CellDataStorage();
```

Comment:
I doesn't hurt, but there's also no point in it here. Virtual functions are useful if anyone is going to store a pointer-to-base, but people only do that if there is polymorphism, which there isn't in this class.

My philosophy these days is that "doing things just because they can be done" is almost always a poor strategy because it makes people think about the *reasons* for these design decisions and wonder whether there are parts of the design they don't understand. Better go with simple.